src/FSharpx.Collections.Experimental/SkewBinomialHeap.fs

```diff
@@ -0,0 +1,307 @@
+
+[<AutoOpen>]
+module private Helper =
+    let inline xor a b = (a && not b) || (not a && b)
+
+module private SBHTree =
+    let item (Node (x, _, _)) = x
+
+    let link descending (Node (x, auxX, childrenX) as treeX) (Node (y, auxY, childrenY) as treeY) =
+        if xor (x <= y) descending
+            then Node (x, auxX, treeY::childrenX)
+            else Node (y, auxY, treeX::childrenY)
+
+    let skewLink descending v treeX treeY =
+        let (Node (w, aux, children)) = link descending treeX treeY
+        if xor (v <= w) descending
+            then Node (v, w::aux, children)
+            else Node (w, v::aux, children)
+
+module private SBHTreeRoot =
+    open SBHTree
+
+    let inline getTree (Root (_, tree)) = tree
+
+    let insert descending value = function
+        | Root(rank1, tree1)::Root(rank2, tree2)::roots when rank1 = rank2 -> Root(rank1 + 1, skewLink descending value tree1 tree2)::roots
+        | roots                                                            -> Root(0, Node (value, [], []))::roots
+
+    let rec insRoot descending (Root(rank, tree) as root) roots =
+        match roots with
+        | []                                  -> [root]
+        | Root(rank', _)::_ when rank < rank' -> root::roots
+        | Root(_, tree')::roots'              -> insRoot descending (Root(rank + 1, link descending tree tree')) roots'
+
+    let rec mergeRoots descending roots1 roots2 =
+        match roots1, roots2 with
+        | roots, [] | [], roots-> roots
+        | Root(rankX, treeX)::roots1', Root(rankY, _)::_ when rankX < rankY -> Root(rankX, treeX)::mergeRoots descending roots1' roots2
+        | Root(rankX, _)::_, Root(rankY, treeY)::roots2' when rankX > rankY -> Root(rankY, treeY)::mergeRoots descending roots1 roots2'
+        | Root(rank, treeX)::roots1', Root(_, treeY)::roots2'               -> insRoot descending (Root(rank + 1, link descending treeX treeY)) (mergeRoots descending roots1' roots2')
+
+    let normalize descending = function
+        | [] -> []
+        | root::roots -> insRoot descending root roots
+
+    let rec extractMinRoot descending = function
+        | [p] -> (p, [])
+        | (Root(_, tree) as root)::roots ->
+            let (Root(_, tree') as root', roots') = extractMinRoot descending roots
+            if xor (SBHTree.item tree <= SBHTree.item tree') descending
+                then root, roots
+                else root', root::roots'
+        | [] -> failwith "This should never happen"
+        
+    let rec findMinRootItem descending = function
+        | [Root(_, node)]                       -> item node
+        | Root(_, node)::roots'                 -> 
+            let this = item node
+            let other = findMinRootItem descending roots'
+            if xor (this <= other) descending
+                then this
+                else other
+        | []                                    -> failwith "This should never happen"
```

Comment:
I'm thinking of creating an exception type for this kind of problems. But still I don't know was message to put in. I don't know a best practice to solve that, if I use a very explicit message explaining the failure I could end up disclosing information (to the user) about the insights of the code and how it fails. If you have any suggestion?